lib/spack/spack/spec.py

```diff
@@ -3514,76 +3656,72 @@ def satisfies(self, other, deps=True, strict=False):
-            var_strict = True
-        if not self.variants.satisfies(other.variants, strict=var_strict):
+        if not self.variants.satisfies(other.variants):
             return False
 
         # Architecture satisfaction is currently just string equality.
         # If not strict, None means unconstrained.
         if self.architecture and other.architecture:
-            if not self.architecture.satisfies(other.architecture, strict):
+            if not self.architecture.satisfies(other.architecture):
                 return False
-        elif strict and (other.architecture and not self.architecture):
+        elif other.architecture and not self.architecture:
             return False
 
-        if not self.compiler_flags.satisfies(other.compiler_flags, strict=strict):
+        if not self.compiler_flags.satisfies(other.compiler_flags):
             return False
 
         # If we need to descend into dependencies, do it, otherwise we're done.
         if deps:
-            deps_strict = strict
-            if self._concrete and not other.name:
-                # We're dealing with existing specs
-                deps_strict = True
-            return self.satisfies_dependencies(other, strict=deps_strict)
+            return self._satisfies_dependencies(other)
         else:
             return True
 
-    def satisfies_dependencies(self, other, strict=False):
-        """
-        This checks constraints on common dependencies against each other.
-        """
+    def _satisfies_dependencies(self, other):
         other = self._autospec(other)
 
         # If there are no constraints to satisfy, we're done.
         if not other._dependencies:
             return True
 
-        if strict:
-            # if we have no dependencies, we can't satisfy any constraints.
-            if not self._dependencies:
-                return False
-
-            # use list to prevent double-iteration
-            selfdeps = list(self.traverse(root=False))
-            otherdeps = list(other.traverse(root=False))
-            if not all(any(d.satisfies(dep, strict=True) for d in selfdeps) for dep in otherdeps):
-                return False
+        # If we have no dependencies, we can't satisfy any constraints.
+        if not self._dependencies:
+            return False
 
-        elif not self._dependencies:
-            # if not strict, this spec *could* eventually satisfy the
-            # constraints on other.
-            return True
+        # use list to prevent double-iteration
+        selfdeps = list(self.traverse(root=False))
+        otherdeps = list(other.traverse(root=False))
+        if not all(any(d.satisfies(dep) for d in selfdeps) for dep in otherdeps):
+            return False
 
         # Handle first-order constraints directly
         for name in self.common_dependencies(other):
             if not self[name].satisfies(other[name], deps=False):
                 return False
 
+        if self.concrete:
```

Comment:
This is dead code? You already checked that all rhs satisfy any lhs

Reply:
No, this is the part that fixed #35597 originally.

Reply:
I'm not a fan of this implementation to be honest. Why don't you replace it with a one-liner and leave the provider business to `satisfies`?

```
def satisfies_deps(self, other):
    return all(any(lhs.satisfies(rhs) for lhs in self.traverse(root=False)) for rhs in other.traverse(root=False))
```

Reply:
Fine with me, I tried to maintain most of the previous logic in this PR, but I agree that there are simpler implementations - in particular now that we split the `intersects` vs. `satisfies` semantic.

Reply:
imo it's less likely to be incorrect, since this is pretty much the property we need (assuming the graph structure is irrelevant for the abstract rhs...)

Reply:
Ok, done in cf3a79a2e98a9881382a4d898125d5e9389ef9c0. We have to revisit how we handle structured DAGs as soon as Spack handles "separate concretization of build dependencies" and we can use syntax like:
```
hd5 %( cmake ^zlib~shared )
```
to have a non flat representation of an abstract DAG.